### PR TITLE
CDAP-13274 Emit FLL from GCS source and sink

### PR DESCRIPTION
**Pull request by rohitsinha54**
Emit FLL from GCS source and sink

Operation:
image

Source Field
image

This PR also fixes https://issues.cask.co/browse/CDAP-14353

### CODE REVIEW

File: src/main/java/co/cask/gcp/gcs/source/AbstractFileBatchSource.java [hunk 1]
```diff
@@ -82,15 +81,17 @@ public void prepareRun(BatchSourceContext context) throws Exception {
       conf.set(entry.getKey(), entry.getValue());
     }
 
-    conf.set(INPUT_REGEX_CONFIG, config.fileRegex);
+    if (config.fileRegex != null) {
+      conf.set(INPUT_REGEX_CONFIG, config.fileRegex);
+    }
 
-    FileInputFormat.setInputDirRecursive(job, config.recursive);
+    FileInputFormat.setInputDirRecursive(job, config.recursive == null ? false : config.recursive);
```
Comment (albertshau) [inline src/main/java/co/cask/gcp/gcs/source/AbstractFileBatchSource.java]: can you add getters to the config that do this logic

File: src/main/java/co/cask/gcp/gcs/source/GCSSource.java [hunk 2]
```diff
@@ -50,6 +60,16 @@ public void prepareRun(BatchSourceContext context) throws Exception {
     super.prepareRun(context);
     config.validate();
     GCPUtils.getProjectId(config.project);
+
+    // record field level lineage information
+    Schema outputSchema = context.getOutputSchema();
+    if (outputSchema != null && outputSchema.getFields() != null && !outputSchema.getFields().isEmpty()) {
+      FieldOperation operation = new FieldReadOperation("Read", "Read from Google Cloud Storage Bucket.",
```
Comment (albertshau) [inline src/main/java/co/cask/gcp/gcs/source/GCSSource.java]: Read from Google Cloud Storage.

File: src/main/java/co/cask/gcp/gcs/source/AbstractFileBatchSource.java [hunk 3]
```diff
@@ -82,15 +81,17 @@ public void prepareRun(BatchSourceContext context) throws Exception {
       conf.set(entry.getKey(), entry.getValue());
     }
 
-    conf.set(INPUT_REGEX_CONFIG, config.fileRegex);
+    if (config.fileRegex != null) {
+      conf.set(INPUT_REGEX_CONFIG, config.fileRegex);
+    }
 
-    FileInputFormat.setInputDirRecursive(job, config.recursive);
+    FileInputFormat.setInputDirRecursive(job, Boolean.TRUE.equals(config.recursive));
```
Comment (albertshau) [inline src/main/java/co/cask/gcp/gcs/source/AbstractFileBatchSource.java]: previous comment was about adding a isRecursive(), shouldIgnoreNonExistingFolders() and useFilenameOnly() methods to the config class that return boolean and not Boolean. Those properties should not be public.

Comment (rohitsinha54) [reply src/main/java/co/cask/gcp/gcs/source/AbstractFileBatchSource.java]: Oh I thought that the intention was not have null checks hanging around. Will add the getters

**Review by albertshau — APPROVED** [review]
lgtm

**Pull request merged**